### ISSUE TITLE
Infinite loop when ``c-extension-no-member`` is activated 

### ISSUE THREAD

**Issue by wieczorek1990** [pylint-dev/pylint#7585]
### Bug description

pylint does not complete linitng.
Warnings are printed when available but the program does not complete.

### Configuration

```ini
[tool.pylint]
disable = [
    "C0103",
    "C0114",
    "C0115",
    "C0116",
    "C0301",
    "C0413",
    "C0415",
    "E1101",
    "R0901",
    "R0902",
    "R0903",
    "R0904",
    "R0911",
    "R0913",
    "R0914",
    "W0212",
    "W0221",
]
ignore = "migrations"
max-line-length = 80
```

### Command used

```shell
PYTHONPATH=app/ pylint --jobs 0 $apps
# same for
PYTHONPATH=app/ pylint $apps
```

### Pylint output

```shell
empty output
```

### Expected behavior

Does not loop forever.

### Pylint version

```shell
pylint 3.0.0-a5
astroid 2.11.7
Python 3.10.7 (main, Sep 15 2022, 01:51:29) [Clang 14.0.0 (clang-1400.0.29.102)]
```

### OS / Environment

macOS 12.6 (21G115)
Alpine Linux v3.16

**Comment by wieczorek1990** [comment]
The problem is with the following code:

I have a test like this:

```
    def test_post_other_integration(self) -> None:
        self.user_credentials.integration = "unknown"

        poster = postings.BexioPoster(self.posting_heading)
        success = poster.post()
        self.assertFalse(success)
 ```
 
 I have Django model like this:
 
 ```
class Integration(django.db.models.TextChoices):
    BEXIO = luca.typing.string_auto("bexio", "Bexio")


class IntegrationField(django.db.models.CharField):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.max_length = 16
        self.choices = Integration.choices

 class UserCredentials(luca.models.TimeStampedModel):
    integration = IntegrationField()
 ```

The typing code is here:

```
import typing


class StringValueTuple(tuple):
    value: str


def auto(
    code: typing.Any, name: str, type_: typing.Type[typing.Any]
) -> typing.Any:
    return type_([code, name])


def string_auto(code: str, name: str) -> StringValueTuple:
    return auto(code, name, StringValueTuple)
```

The test is invalid because it assigns a string to an annotated enum.

pylint is stuck on branch `v3.0.0-a5` in `pylint.checkers.typecheck._similar_names:187`.
I've checked that it does not enter `_node_names` and is somewhere it `functools.singledispatch`.

Debug context:
```
attrname = {str} 'user_credentials'
distance_threshold = {int} 1
max_choices = {int} 1
owner = {Instance} Instance of ledgers.tests.test_postings.PostingsTestCase
  _proxied = {ClassDef} ClassDef.PostingsTestCase(name='PostingsTestCase',\n                          doc=None,\n                          is_dataclass=False,\n                          position=Position(lineno=11, col_offset=0, end_lineno=11, end_col_offset=22),\n                   
possible_names = {list: 0} []
 __len__ = {int} 0
 ```

**Comment by jacobtylerwalls** [comment]
Interesting. Can you reproduce against a stable version like 2.15.3?

**Comment by wieczorek1990** [comment]
Same on that version.

W dniu pt., 7.10.2022 o 18:45 Jacob Walls ***@***.***>
napisał(a):

**Comment by wieczorek1990** [comment]
I've tried to reproduce here:
https://github.com/wieczorek1990/pylintloop
But without any luck.

**Comment by wieczorek1990** [comment]
If I add `I1101` to disabled my next problem with infinite loop is solved.

**Comment by nickdrozd** [comment]
For reference, `I1101` is `c-extension-no-member`:

```
:c-extension-no-member (I1101): *%s %r has no %r member%s, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects.*
  Used when a variable is accessed for non-existent member of C extension. Due
  to unavailability of source static analysis is impossible, but it may be
  performed by introspecting living objects in run-time.
```

I have nothing else to add, except to say that "introspection of living objects" is quite an evocative expression.

**Comment by Pierre-Sassoulas** [comment]
I still cannot reproduce and this is now an old version so I'm going to close to clean up the backlog. Thank you for opening the issue in the first place, please feel free to reopen with a reproducer targeting the latest pylint if the issue still exists.
